File: code_samples/dijkstra/dijkstra.py
"""
Dijkstra's Algorithm implementation in Python.

Dijkstra's algorithm is a graph search algorithm that finds the shortest path
between nodes in a graph. It works by visiting vertices in order of increasing
distance from the source.

Time Complexity: O(V^2) with adjacency matrix, O((V+E)logV) with min-heap
Space Complexity: O(V)
"""

import heapq

def dijkstra(graph, start):
    """
    Find shortest paths from start vertex to all vertices in the graph.
    
    Args:
        graph: Dictionary representing weighted adjacency list of the graph
               {vertex: {neighbor: weight, ...}, ...}
        start: Starting vertex
        
    Returns:
        Dictionary of shortest distances and dictionary of predecessors
    """
    # Initialize distances with infinity for all vertices except start
    distances = {vertex: float('infinity') for vertex in graph}
    distances[start] = 0
    
    # Initialize predecessors
    predecessors = {vertex: None for vertex in graph}
    
    # Priority queue to store vertices to be processed
    # Format: (distance, vertex)
    priority_queue = [(0, start)]
    
    # Set to keep track of vertices already processed
    processed = set()
    
    while priority_queue:
        # Get vertex with minimum distance
        current_distance, current_vertex = heapq.heappop(priority_queue)
        
        # If already processed or found a longer path, skip
        if current_vertex in processed or current_distance > distances[current_vertex]:
            continue
        
        # Mark as processed
        processed.add(current_vertex)
        
        # Check all neighbors of current vertex
        for neighbor, weight in graph[current_vertex].items():
            # Calculate distance to neighbor through current vertex
            distance = current_distance + weight
            
            # If found a shorter path to neighbor
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                predecessors[neighbor] = current_vertex
                heapq.heappush(priority_queue, (distance, neighbor))
    
    return distances, predecessors

def get_shortest_path(predecessors, target):
    """
    Reconstruct shortest path from start to target using predecessors.
    
    Args:
        predecessors: Dictionary of predecessors
        target: Target vertex
        
    Returns:
        List representing the shortest path from start to target
    """
    path = []
    current = target
    
    # Reconstruct path from target to start
    while current is not None:
        path.append(current)
        current = predecessors[current]
    
    # Return reversed path (from start to target)
    return path[::-1]


# Example usage
if __name__ == "__main__":
    # Example graph represented as a weighted adjacency list
    graph = {
        'A': {'B': 4, 'C': 2},
        'B': {'A': 4, 'D': 2, 'E': 3},
        'C': {'A': 2, 'D': 4, 'F': 5},
        'D': {'B': 2, 'C': 4, 'E': 1, 'F': 7},
        'E': {'B': 3, 'D': 1, 'F': 4},
        'F': {'C': 5, 'D': 7, 'E': 4}
    }
    
    start_vertex = 'A'
    target_vertex = 'F'
    
    # Find shortest paths from start vertex
    distances, predecessors = dijkstra(graph, start_vertex)
    
    # Print shortest distances from start vertex
    print(f"Shortest distances from {start_vertex}:")
    for vertex, distance in distances.items():
        print(f"{vertex}: {distance}")
    
    # Print shortest path to target vertex
    shortest_path = get_shortest_path(predecessors, target_vertex)
    print(f"\nShortest path from {start_vertex} to {target_vertex}: {shortest_path}")
    print(f"Distance: {distances[target_vertex]}")

File: code_samples/dijkstra/dijkstra.java
/**
 * Dijkstra's Algorithm implementation in Java.
 *
 * Dijkstra's algorithm is a graph search algorithm that finds the shortest path
 * between nodes in a graph. It works by visiting vertices in order of increasing
 * distance from the source.
 *
 * Time Complexity: O(V^2) with adjacency matrix, O((V+E)logV) with min-heap
 * Space Complexity: O(V)
 */
import java.util.*;

public class Dijkstra {
    
    /**
     * Find shortest paths from start vertex to all vertices in the graph.
     *
     * @param graph Weighted adjacency list of the graph
     * @param start Starting vertex
     * @return Map of shortest distances and map of predecessors
     */
    public static Map<String, Object> dijkstra(Map<String, Map<String, Integer>> graph, String start) {
        // Initialize distances with infinity for all vertices except start
        Map<String, Integer> distances = new HashMap<>();
        for (String vertex : graph.keySet()) {
            distances.put(vertex, Integer.MAX_VALUE);
        }
        distances.put(start, 0);
        
        // Initialize predecessors
        Map<String, String> predecessors = new HashMap<>();
        for (String vertex : graph.keySet()) {
            predecessors.put(vertex, null);
        }
        
        // Priority queue to store vertices to be processed
        // Format: (vertex, distance)
        PriorityQueue<Map.Entry<String, Integer>> priorityQueue = new PriorityQueue<>(
            Comparator.comparing(Map.Entry::getValue)
        );
        priorityQueue.add(new AbstractMap.SimpleEntry<>(start, 0));
        
        // Set to keep track of vertices already processed
        Set<String> processed = new HashSet<>();
        
        while (!priorityQueue.isEmpty()) {
            // Get vertex with minimum distance
            Map.Entry<String, Integer> entry = priorityQueue.poll();
            String currentVertex = entry.getKey();
            int currentDistance = entry.getValue();
            
            // If already processed or found a longer path, skip
            if (processed.contains(currentVertex) || currentDistance > distances.get(currentVertex)) {
                continue;
            }
            
            // Mark as processed
            processed.add(currentVertex);
            
            // Check all neighbors of current vertex
            Map<String, Integer> neighbors = graph.get(currentVertex);
            for (Map.Entry<String, Integer> neighborEntry : neighbors.entrySet()) {
                String neighbor = neighborEntry.getKey();
                int weight = neighborEntry.getValue();
                
                // Calculate distance to neighbor through current vertex
                int distance = currentDistance + weight;
                
                // If found a shorter path to neighbor
                if (distance < distances.get(neighbor)) {
                    distances.put(neighbor, distance);
                    predecessors.put(neighbor, currentVertex);
                    priorityQueue.add(new AbstractMap.SimpleEntry<>(neighbor, distance));
                }
            }
        }
        
        // Return both distances and predecessors
        Map<String, Object> result = new HashMap<>();
        result.put("distances", distances);
        result.put("predecessors", predecessors);
        
        return result;
    }
    
    /**
     * Reconstruct shortest path from start to target using predecessors.
     *
     * @param predecessors Map of predecessors
     * @param target Target vertex
     * @return List representing the shortest path from start to target
     */
    public static List<String> getShortestPath(Map<String, String> predecessors, String target) {
        List<String> path = new ArrayList<>();
        String current = target;
        
        // Reconstruct path from target to start
        while (current != null) {
            path.add(current);
            current = predecessors.get(current);
        }
        
        // Return reversed path (from start to target)
        Collections.reverse(path);
        return path;
    }
    
    // Example usage
    public static void main(String[] args) {
        // Example graph represented as a weighted adjacency list
        Map<String, Map<String, Integer>> graph = new HashMap<>();
        
        graph.put("A", new HashMap<>());
        graph.get("A").put("B", 4);
        graph.get("A").put("C", 2);
        
        graph.put("B", new HashMap<>());
        graph.get("B").put("A", 4);
        graph.get("B").put("D", 2);
        graph.get("B").put("E", 3);
        
        graph.put("C", new HashMap<>());
        graph.get("C").put("A", 2);
        graph.get("C").put("D", 4);
        graph.get("C").put("F", 5);
        
        graph.put("D", new HashMap<>());
        graph.get("D").put("B", 2);
        graph.get("D").put("C", 4);
        graph.get("D").put("E", 1);
        graph.get("D").put("F", 7);
        
        graph.put("E", new HashMap<>());
        graph.get("E").put("B", 3);
        graph.get("E").put("D", 1);
        graph.get("E").put("F", 4);
        
        graph.put("F", new HashMap<>());
        graph.get("F").put("C", 5);
        graph.get("F").put("D", 7);
        graph.get("F").put("E", 4);
        
        String startVertex = "A";
        String targetVertex = "F";
        
        // Find shortest paths from start vertex
        Map<String, Object> result = dijkstra(graph, startVertex);
        Map<String, Integer> distances = (Map<String, Integer>) result.get("distances");
        Map<String, String> predecessors = (Map<String, String>) result.get("predecessors");
        
        // Print shortest distances from start vertex
        System.out.println("Shortest distances from " + startVertex + ":");
        for (Map.Entry<String, Integer> entry : distances.entrySet()) {
            System.out.println(entry.getKey() + ": " + entry.getValue());
        }
        
        // Print shortest path to target vertex
        List<String> shortestPath = getShortestPath(predecessors, targetVertex);
        System.out.println("\nShortest path from " + startVertex + " to " + targetVertex + ": " + shortestPath);
        System.out.println("Distance: " + distances.get(targetVertex));
    }
}
